File: 71-simplify-path/simplify-path.ts
function simplifyPath(path: string): string {
     const components = path.split("/");
    const st = [];

    for (const comp of components) {
        if (comp === "" || comp === ".") {
            continue;
        }

        if (comp === "..") {
            if (st.length) {
                st.pop();
            }
        } else {
            st.push(comp);
        }
    }

    return "/" + st.join("/");    
};

File: 71-simplify-path/simplify-path.java
class Solution {
   public String simplifyPath(String path) {
    ArrayDeque<String> st = new ArrayDeque<>();

    for (String part : path.split("/")) {
        if (part.isEmpty() || part.equals(".")) {
            continue;
        }

        if (part.equals("..")) {
            if (!st.isEmpty()) {
                st.pop();
            }
        } else {
            st.push(part);
        }
    }

    StringBuilder ans = new StringBuilder();

    while (!st.isEmpty()) {
        ans.append("/").append(st.removeLast());
    }

    return ans.length() == 0 ? "/" : ans.toString();
}
}